src/psyclone/transformations.py

```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+            if node.abs_position <= taskloop.abs_position:
+                continue
+            # Ignore our children
+            anc = node.ancestor(OMPTaskloopDirective)
+            if anc is taskloop:
+                continue
+            node_vars = None
+            # Special case:
+            # 1. If we have a different parent parallel node, then our parent
+            # parallel node is our dependency
+            if node.ancestor(OMPParallelDirective) is not parent_parallel:
+                return parent_parallel
+
+            if (isinstance(node, OMPTaskwaitDirective) and
+                    (taskloop.ancestor(OMPSingleDirective) is
+                     node.ancestor(OMPSingleDirective)) and
+                    taskloop.ancestor(OMPSingleDirective) is not None and
+                    taskloop.ancestor(OMPSingleDirective).nowait is False):
+                # If we find a taskwait barrier inside the same
+                # OMPSingleDirective and that OMPSingleDirective has a no
+                # nowait clause, then it acts as a barrier for dependencies
+                # as well
+                node_vars = taskloop_vars
+            elif not isinstance(node, (OMPSerialDirective,
+                                       OMPParallelDirective)):
+                # For all our other node types that cannot be our parents,
+                # they use their own signatures
+                node_vars = VariablesAccessInfo()
+                for child in node.walk(nodes.Node):
+                    if child is not node and not isinstance(child,
+                                                            (Schedule, Loop)):
+                        node_vars.merge(VariablesAccessInfo(child))
+            else:
+                # If we find node types that could be our parents (are are
+                # known to not be valid options from previous checks) then
+                # we set node_vars to empty
+                node_vars = VariablesAccessInfo()
+            node_signatures = node_vars.all_signatures
+            # Once we have the node's variable accesses, check for collisions
+            for sig1 in taskloop_signatures:
+                # If this signature is not in the node signatures then continue
+                if sig1 not in node_signatures:
+                    continue
+                access1 = taskloop_vars[sig1]
+                access2 = node_vars[sig1]
+                # If both read only, we can ignore this signature
+                # Otherwise, one of themwrites so return this node as
+                # we have a WaR or RaW dependency
+                if access1.is_written() or access2.is_written():
+                    # If we have a different parent serial node, and
+                    # parent_single is not None then our parent_single is our
+                    # dependency, otherwise this node is the dependency
+                    if (taskloop.ancestor(OMPSerialDirective) is not
```

Comment:
Use cached result for `taskloop.ancestor(OMPSerialDirective)`?

Reply:
I previously only computed `taskloop.ancestor(OMPSingleDirective)` - the difference here is there is another subclass of `OMPSerialDirective` which is legal to contain these nodes (`OMPMasterDirective`), but does not allow some dependency patterns between regions to be satisfiable.